File: src/main/java/cn/xor7/iseeyou/ConfigData.java
package cn.xor7.iseeyou;

import org.bukkit.entity.Player;

import javax.annotation.Nullable;
import java.util.HashSet;
import java.util.Set;

/**
 * @author MC_XiaoHei
 */
public class ConfigData {
    @Nullable
    public String isConfigValid() {
        if (!"name".equals(checkBy) && !"uuid".equals(checkBy)) {
            return "invalid checkBy value!";
        }
        if (!"blacklist".equals(recordMode) && !"whitelist".equals(recordMode)) {
            return "invalid recordMode value!";
        }
        return null;
    }

    public void setConfig() {
        if ("blacklist".equals(recordMode)) {
            if (blacklist == null) {
                blacklist = new HashSet<>();
            }
        } else if (whitelist == null) {
            whitelist = new HashSet<>();
        }
    }

    public boolean shouldRecordPlayer(Player player) {
        if ("blacklist".equals(recordMode)) {
            return !containsPlayer(player, blacklist);
        } else {
            return containsPlayer(player, whitelist);
        }
    }

    private boolean containsPlayer(Player player, Set<String> list) {
        if ("name".equals(checkBy)) {
            return list.contains(player.getName());
        } else {
            return list.contains(player.getUniqueId().toString());
        }
    }

    public static final class HighSpeedPauseConfig {
        public Boolean enable = false;
        public Double threshold = 20.00;
    }

    public HighSpeedPauseConfig pauseRecordingOnHighSpeed = new HighSpeedPauseConfig();
    public Boolean deleteTmpFileOnLoad = true;
    public Boolean pauseInsteadOfStopRecordingOnPlayerQuit = false;
    public String recordMode = "blacklist";
    public String checkBy = "name";
    public String recordPath = "replay/player/${name}@${uuid}";
    public Set<String> blacklist;
    public Set<String> whitelist;
}

File: src/main/java/cn/xor7/iseeyou/ConfigData.kt
package cn.xor7.iseeyou

import org.bukkit.entity.Player

data class ConfigData(
    var pauseRecordingOnHighSpeed: HighSpeedPauseConfig = HighSpeedPauseConfig(),
    var deleteTmpFileOnLoad: Boolean = true,
    var pauseInsteadOfStopRecordingOnPlayerQuit: Boolean = false,
    var filter: FilterConfig = FilterConfig(),
    var recordPath: String = "replay/player/\${name}@\${uuid}",
    var clearOutdatedRecordFile: OutdatedRecordRetentionConfig = OutdatedRecordRetentionConfig(),
    var recordSuspiciousPlayer: RecordSuspiciousPlayerConfig = RecordSuspiciousPlayerConfig(),
    // var asyncSave: Boolean = false,
) {
    fun isConfigValid(): String? {
        if ("name" != filter.checkBy && "uuid" != filter.checkBy) {
            return "invalid checkBy value!"
        }
        if ("blacklist" != filter.recordMode && "whitelist" != filter.recordMode) {
            return "invalid recordMode value!"
        }
        return null
    }

    fun setConfig() {
        if ("blacklist" == filter.recordMode) {
            if (filter.blacklist == null) {
                filter.blacklist = mutableSetOf()
            }
        } else if (filter.whitelist == null) {
            filter.whitelist = mutableSetOf()
        }
    }

    fun shouldRecordPlayer(player: Player): Boolean {
        return if ("blacklist" == filter.recordMode) {
            !containsPlayer(player, filter.blacklist!!)
        } else {
            containsPlayer(player, filter.whitelist!!)
        }
    }

    private fun containsPlayer(player: Player, list: Set<String>): Boolean {
        return if ("name" == filter.checkBy) {
            list.contains(player.name)
        } else {
            list.contains(player.uniqueId.toString())
        }
    }
}

data class HighSpeedPauseConfig(
    var enabled: Boolean = false,
    var threshold: Double = 20.00,
)

data class OutdatedRecordRetentionConfig(
    var enabled: Boolean = false,
    var days: Int = 7,
)

data class RecordSuspiciousPlayerConfig(
    var enableThemisIntegration: Boolean = false,
    var enableMatrixIntegration: Boolean = false,
    var recordMinutes: Long = 5,
    var recordPath: String = "replay/suspicious/Themis/\${name}@\${uuid}",
)

data class FilterConfig(
    var checkBy: String = "name",
    var recordMode: String = "blacklist",
    var blacklist: Set<String>? = null,
    var whitelist: Set<String>? = null,
)
